cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/artifact/ArtifactStore.java

```diff
@@ -571,13 +571,81 @@ public boolean apply(PluginClass input) {
   }
 
   /**
-   * Update artifact properties using an update function. Functions will receive an immutable map.
+   * Get all plugin classes of the given type and name that extend the given parent artifact.
+   * Results are returned as a map from plugin artifact to plugins in that artifact.
    *
-   * @param artifactId the id of the artifact to add
-   * @param updateFunction the function used to update existing properties
-   * @throws ArtifactNotFoundException if the artifact does not exist
-   * @throws IOException if there was an exception writing the properties to the metastore
+   * @param parentArtifactRange the parent artifact range to find plugins for
+   * @param type the type of plugin to look for
+   * @param name the name of the plugin to look for
+   * @return an unmodifiable map of plugin artifact to plugin classes of the given type and name, accessible by the
+   *         given artifact. The map will never be null, and will never be empty.
+   * @throws PluginNotExistsException if no plugin with the given type and name exists in the namespace
+   * @throws IOException if there was an exception reading metadata from the metastore
    */
+  public SortedMap<ArtifactDescriptor, PluginClass> getPluginClasses(final NamespaceId namespace,
+                                                                     final ArtifactRange parentArtifactRange,
+                                                                     final String type, final String name)
+    throws IOException, ArtifactNotFoundException, PluginNotExistsException {
+    final List<ArtifactDetail> artifactDetails = getArtifacts(parentArtifactRange);
```

Comment:
shouldn't do this outside of the transaction. You can refactor getArtifacts() so that all of the logic inside the transaction is in a separate method that getArtifacts() and this method both call